Subtask-related features

MangoCubes opened the pull request:
Found out you can't set parent task in this web GUI, so I added it. The button is not clear, but at least you can select the parent task. Kinda fixes #124.

https://github.com/etesync/etesync-web/assets/10383115/74c529df-549c-464f-86dc-9947b1d01005

MangoCubes committed:
Tasks edit: Simplified task selection

barathrm commented:
This is awesome! :) Would be cool to be able to create a child task with some more UX (add child, or a context icon), but even just adding this would be new functionality and improvement :rocket: 

fyi @tasn

barathrm reviewed (commented):
I'm not a webdev so I can't comment on that without more work, but I'd bundle your commits into logical changes and try to avoid commits names like [Setting task parent is now working](https://github.com/etesync/etesync-web/pull/262/commits/29fe78b0411eadcbd76f4f5d7bff45ce56adc1a0). Those are great during development, but imo a commit history is better off containing "chunk" of work so that it's easier to test what causes a bug or easier to revert something by just revert 1-2 commits.

So since I'm not a webdev it's not immediately obvious to me which commits you might want to squash, but if everything is new and there are no single stand-alone changes that could "live on their own" then maybe just one commit? :shrug:

MangoCubes commented:
In terms of what users would see, yeah, I don't exactly see how any of the changes could 'live on their own' either. It's one feature and all commits are supporting one another. But some of these 'supporting features' can sort of live on their own, and it would be something like this:

[Task selector](https://github.com/etesync/etesync-web/pull/262/commits/312a0f2cf22fb3e8628f9fa0e320c24c5791940a)~[Parent task added into TaskEdit](https://github.com/etesync/etesync-web/pull/262/commits/4b467790f428632dd29e49e0c9102483e1ed7dc7): Adding dialog window itself, edits made via dialog windows doesn't actually change anything
[Setting task parent is now working](https://github.com/etesync/etesync-web/pull/262/commits/29fe78b0411eadcbd76f4f5d7bff45ce56adc1a0), [Task can no longer be child of its children](https://github.com/etesync/etesync-web/pull/262/commits/2b22c9cd810b8ef07f08525581c23b481f064358): Critical commits that makes the changes in the dialog goes through
[Simplified selection step](https://github.com/etesync/etesync-web/pull/262/commits/bda98e1db28e88c54d09445d8960d9a35ec1cf4b), [Selection changes field and closes the dialog](https://github.com/etesync/etesync-web/pull/262/commits/fbeb6bfce8215d4542317434680de5d6f0d92b69), [Setting default parent if it got one](https://github.com/etesync/etesync-web/pull/262/commits/06a455c20320f9ed860ddd5a605818b8f48cb15e): Making it easier for users to use the dialog

Oh, and looking back to it, I do agree that commit's message is crap. It was meant to be that one critical commit where the changes user make actually gets applied and it's no longer dry runs only. But yeah, should have been something like "Change in task parent is applied and gets sent to the server".

So regarding these two (squashing and changing commit messages), should I do them myself, and get back once that's done?

barathrm commented:
Nice, those three groupings sound like they might make sense to bundle things into, so that'd end up as 3 commits after squashing/interactive rebasing. 

> Feedback on how I grouped the commits would be great; I have some more features I want to see in EteSync, and need some samples on what you're looking for.

Note that I'm just a drive-by commenter, not a contributor on this repo :D I just package etesync for some platforms and am a (non-web) dev in general, but ultimately it's up to the repo owner (@tasn). 

When in doubt, it's always good to look for existing patterns: https://github.com/etesync/etesync-web/commits/master

It looks like most commits add a single "feature" or atomic change, and that commit message titles follow a pattern of `<area/topic>: <short description of change>`.

> So regarding these two (squashing and changing commit messages), should I do them myself, and get back once that's done?

Yeah I'd just do an interactive rebase (`git rebase -i HEAD~13` or something if you're a terminal user) and squash and edit, then force-push the changed branch, as I think bundling this feature into fewer commits is something we'd want anyways. :)

MangoCubes committed:
Tasks edit: Added parent task selector

MangoCubes committed:
Tasks edit: Basic functionality for task selector

MangoCubes commented:
I'm pretty sure it went as intended; Let's hope tasn sees this.

MangoCubes committed:
Tasks edit: Added toggle switch for hidden and completed tasks

MangoCubes committed:
Merge pull request #2 from MangoCubes/subtask

Tasks edit: Added toggle switch for hidden and completed tasks

MangoCubes committed:
Helper functions: Functions updated to allow multiple edits/deletes

MangoCubes committed:
Tasks edit: Added components for subtask list in the edit screen

MangoCubes committed:
Tasks edit: Implemented core functionality of the subtask creation feature

MangoCubes committed:
Tasks edit: Added supporting features (Press enter to create new subtask, no empty task, recursive task delete)

MangoCubes committed:
Tasks: Added a way to view tasks with missing parent

MangoCubes committed:
Tasks edit: Merged parent selector feature into subtask list feature

MangoCubes committed:
Tasks edit: Fixing layout for delete confirmation dialog

MangoCubes committed:
Merge branch 'create-subtask'

MangoCubes commented:
Perhaps I am going a bit too ahead with this, but I have added subtask list in the edit page. This demo shows the feature I have implemented.

https://github.com/etesync/etesync-web/assets/10383115/d65026cd-37ef-4e22-9122-5c1c8b562712

tasn commented:
Whoops, I missed this! Going to look at reviewing this on the weekend. Thanks a lot for this, and thanks @barathrm for pinging me about it!

arc-des commented:
> Whoops, I missed this! Going to look at reviewing this on the weekend. Thanks a lot for this, and thanks @barathrm for pinging me about it!

Hi! Just checking if this is still on the roadmap? I don't seem able to add subtasks to tasks (or assign tasks to a parent task) using the web UI - as in #252 

Thanks so much for the web version by the way, it's so useful!
